apollo-router/src/plugins/progressive_override/mod.rs

```diff
@@ -0,0 +1,273 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use apollo_compiler::schema::ExtendedType;
+use apollo_compiler::Schema;
+use dashmap::DashMap;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use sha2::Digest;
+use sha2::Sha256;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use self::layers::query_analysis::ParsedDocument;
+use self::visitor::OverrideLabelVisitor;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+use crate::services::*;
+use crate::spec;
+use crate::spec::query::traverse;
+
+pub(crate) mod visitor;
+pub(crate) const UNRESOLVED_LABELS_KEY: &str = "apollo_override::unresolved_labels";
+pub(crate) const LABELS_TO_OVERRIDE_KEY: &str = "apollo_override::labels_to_override";
+
+pub(crate) const JOIN_FIELD_DIRECTIVE_NAME: &str = "join__field";
+pub(crate) const JOIN_SPEC_BASE_URL: &str = "https://specs.apollo.dev/join";
+pub(crate) const JOIN_SPEC_VERSION_RANGE: &str = ">=0.4.0, <=0.4.0";
+pub(crate) const OVERRIDE_LABEL_ARG_NAME: &str = "overrideLabel";
+
+/// Configuration for the progressive override plugin
+#[derive(Debug, Default, Deserialize, JsonSchema)]
+pub(crate) struct Config {}
+
+pub(crate) struct ProgressiveOverridePlugin {
+    enabled: bool,
+    schema: Schema,
+    labels_from_schema: LabelsFromSchema,
+    // We have to visit each operation to find out which labels from the schema
+    // are relevant for any given operation. This allows us to minimize the
+    // number of labels we ultimately send to the query planner. Since these
+    // labels are a component of the query plan cache key, it's important we
+    // don't "overprovide" any labels, since doing so can explode the number of
+    // cache entries per operation.
+    labels_per_operation_cache: Arc<DashMap<String, Vec<Arc<String>>>>,
+}
+
+type LabelsFromSchema = (HashMap<Arc<String>, Arc<f64>>, HashSet<Arc<String>>);
```

Comment:
```suggestion
type LabelsFromSchema = (Arc<HashMap<Arc<String>>, Arc<f64>>, Arc<HashSet<Arc<String>>>);
```
more refcounting, because `HashMap` and `HashSet` can make a lot of allocations for their internal accounting